### DISCUSSION TITLE
Opa constraints (enforce Lifecycle lable to compare with project I'd )

### DISCUSSION THREAD

**Discussion by rakesh0022** [open-policy-agent/community#234] [Community]
<img src="https://user-images.githubusercontent.com/92107358/178128739-bccd701a-0503-4de2-9e5b-cc51bf0a27b7.jpg"/>

I am trying to write a policy to check if lifecycle label is any of the value of {dev, qa ,cert-cde, prod-cde,svcs,cert-ncde,prod-ncde}

i have attached the snippet of the code, i am using
The logic is working when I entered lifecycle label as prod or dev apart for other lifecycle labels value the conditions is not getting executed. can any one will able to assist?

**Comment by srenatus** [comment]
The code snippet looks like you're nesting rules, which is not allowed.

> lifecycle label is any of the value of {dev, qa ,cert-cde, prod-cde,svcs,cert-ncde,prod-ncde}

That could be done by something like
```rego
label in {"dev", "qa", "cert-cde" } # etc
```
which is only available if you also import `future.keywords.in`.

**Reply by rakesh0022** [reply]
yes, but I need lifecycle label should match with the environment character , below is example what I need to achieve with the code
for ex ample- if google project name is - nap-noingopd-ool-lz-test   so the character no 11 which is d in this case, which i achieved using 
environment character := substring(projName, 11 ,) so if d is environment character then lifecycle label would be dev, and if environment character is q then lifecycle label would be qa so likewise i need .for   {"dev":"d",  "qa":"q" , "cert-cde":"c",  "prod-cde":"p", "svcs":"p",  
 "cert-ncde":"s","prod-ncde":"g"}

**Reply by anderseknert** [reply]
Hi @rakesh0022 👋 If you can share what you have on the [Rego Playground](https://play.openpolicyagent.org/) (use the Publish button to do so). that would be helpful to people trying to answer.

**Reply by rakesh0022** [reply]
The above code i m running on playground it's running successfully because playground is using latest version, but my local rego version is 0.29 on which I need to execute the code

**Reply by rakesh0022** [reply]
Is there any way I can use older version rego playground

**Reply by srenatus** [reply]
Anything that works in 0.29.0 works with 0.42.1, just not the other way around. So you can still use the playground to share an example. There are no "older version playgrounds" around.

Alternatively, you could make the example reproducible by providing data.json, input.json and policy.rego files that work with `opa eval`. However, as I said, the playground should do the trick.

**Reply by rakesh0022** [reply]
https://play.openpolicyagent.org/p/YWJZPyvEnE
i am trying use the above code on 0.29 version, but this not working.

**Reply by rakesh0022** [reply]
but it works on 0.42 version on playground

**Reply by srenatus** [reply]
Try `=` instead of `:=` in the toplevel rules: https://play.openpolicyagent.org/p/CSmz7zidsB. _But really_, please do try to upgrade. 0.29.0 is quite ancient. 😉

**Reply by rakesh0022** [reply]
no its not working either by changing = instead of :=

**Reply by srenatus** [reply]
Can you please elaborate "not working"? Does it fail to parse, or is the evaluation result not what you expected?

**Reply by rakesh0022** [reply]
the evaluation result is not what I expected because as per this - https://play.openpolicyagent.org/p/2tiWn10O1J  playground result policy validation is showing false, 
however when I executed the same code on 0.29.3 version it does not showed any error and  i am output as  as "no violations found"

**Reply by rakesh0022** [reply]
it should give an violations saying lifecycle label svcs has violations

**Reply by rakesh0022** [reply]
label_is_valid(label_key, label_value_pattern, asset) {
	resource_labels = {"dev":"d" , "qa":"q" , "cert-cde":"c" , "prod-cde":"p" , "svcs":"p" , "cert-ncde":"s" , "prod-ncde":"g" }
	projName = asset.resource.data.name                           # Google project name(cid-noinpoid-plf-ppic-devqa), getting from terraform Plan 
	environment_character = substring(projName, 11, 1)    # (d will be environment character)       
	lifecycle = asset.resource.data.labels["lifecycle"]             # (passing "svpc" as lifecycle label from main.tf file )            
 	policyvalidation := {    
         environment_character == resource_labels[lifecycle]      #(so d == svpc) it should give violation 

    }
}

this is actual function i have written and trying execute this on 0.29.3 rego version, is not throwing any error, but output I am not getting what is expected

**Reply by rakesh0022** [reply]
after hash # those are statements for the understanding

**Reply by rakesh0022** [reply]
any suggestions ?
